snowpack/src/build/optimize.ts

```diff
@@ -0,0 +1,538 @@
+  }
+  throw new Error('UNEXPECTED ENTRYPOINTS: ' + entrypoints);
+}
+
+async function processEntrypoints(
+  originalEntrypointValue: OptimizeOptions['entrypoints'],
+  entrypointFiles: string[],
+  buildDirectoryLoc: string,
+  baseUrl: string,
+) {
+  let htmlEntrypoints: null | ScannedHtmlEntrypoint[];
+  let bundleEntrypoints: string[];
+  // If entrypoints are JS:
+  if (entrypointFiles.every((f) => f.endsWith('.js'))) {
+    htmlEntrypoints = null;
+    bundleEntrypoints = entrypointFiles;
+    return {htmlEntrypoints, bundleEntrypoints};
+  }
+  // If entrypoints are HTML:
+  if (entrypointFiles.every((f) => f.endsWith('.html'))) {
+    const rawHtmlEntrypoints = await scanHtmlEntrypoints(entrypointFiles);
+    htmlEntrypoints = rawHtmlEntrypoints.filter(isTruthy);
+    if (
+      originalEntrypointValue !== 'auto' &&
+      rawHtmlEntrypoints.length !== htmlEntrypoints.length
+    ) {
+      throw new Error('INVALID HTML ENTRYPOINTS: ' + originalEntrypointValue);
+    }
+    htmlEntrypoints.forEach((val) => extractBaseUrl(val, baseUrl));
+    htmlEntrypoints.forEach(extractInlineScripts);
+    bundleEntrypoints = Array.from(
+      htmlEntrypoints.reduce((all, val) => {
+        val.getLinks('stylesheet').each((_, elem) => {
+          if (!elem.attribs.href || isRemoteSpecifier(elem.attribs.href)) {
+            return;
+          }
+          const resolvedCSS =
+            elem.attribs.href[0] === '/'
+              ? path.resolve(buildDirectoryLoc, removeLeadingSlash(elem.attribs.href))
+              : path.resolve(val.file, '..', elem.attribs.href);
+          all.add(resolvedCSS);
+        });
+        val.getScripts().each((_, elem) => {
+          if (!elem.attribs.src || isRemoteSpecifier(elem.attribs.src)) {
+            return;
+          }
+          const resolvedJS =
+            elem.attribs.src[0] === '/'
+              ? path.join(buildDirectoryLoc, removeLeadingSlash(elem.attribs.src))
+              : path.join(val.file, '..', elem.attribs.src);
+          all.add(resolvedJS);
+        });
+        return all;
+      }, new Set<string>()),
+    );
+    return {htmlEntrypoints, bundleEntrypoints};
+  }
+  // If entrypoints are mixed or neither, throw an error.
+  throw new Error('MIXED ENTRYPOINTS: ' + entrypointFiles);
+}
+
+async function bundleAnalyzeBuildDirectory(
+  bundleEntrypoints: string[],
+  config: SnowpackConfig,
+  esbuildService: esbuild.Service,
+): Promise<{manifest: ESBuildMetaManifest; outputFiles: esbuild.OutputFile[]}> {
+  const {outputFiles, warnings} = await esbuildService.build({
+    entryPoints: bundleEntrypoints,
+    outdir: FAKE_BUILD_DIRECTORY,
+    outbase: config.buildOptions.out,
+    write: false,
+    bundle: true,
+    splitting: true,
+    format: 'esm',
+    platform: 'browser',
+    metafile: path.join(config.buildOptions.out, 'build-manifest.json'),
+    publicPath: config.buildOptions.baseUrl,
+    minify: config.experiments.optimize!.minify,
```

Comment:
Just a note: I’m seeing `minify` work for JS and CSS, but not HTML. Is that intentional?

Reply:
intentional for this PR, but agreed that we should tackle eventually. My hope is that esbuild supports HTML one day, and then we'd get it for free. But, until then we may want to ship some popular html minifier.

I couldn't find one on NPM that I really liked, @drwpow any favorites that you've used in the past?

Reply:
Not really. Fine for waiting to see if esbuild supports it as you said